src/distributions/integer.rs

```diff
@@ -109,53 +107,67 @@ impl_nzint!(NonZeroU64, NonZeroU64::new);
+                        rng.fill_bytes(b_ptr);
+                    }
+                    vec.to_le()
+                }
             }
-        }
-    )+};
+        )+
+    };
+}
 
-    ($bits:expr,) => {};
-    ($bits:expr, $ty:ty, $($ty_more:ty,)*) => {
-        simd_impl!($bits, $($ty_more,)*);
+#[cfg(feature = "simd_support")]
+#[rustfmt::skip]
+packed_simd_types_impl!(
+    u8x2, i8x2,
+    u8x4, i8x4, u16x2, i16x2,
+    u8x8, i8x8, u16x4, i16x4, u32x2, i32x2,
+    u8x16, i8x16, u16x8, i16x8, u32x4, i32x4, u64x2, i64x2, u128x1, i128x1,
+    u8x32, i8x32, u16x16, i16x16, u32x8, i32x8, u64x4, i64x4, u128x2, i128x2,
+    u8x64, i8x64, u16x32, i16x32, u32x16, i32x16, u64x8, i64x8, u128x4, i128x4
+);
 
-        impl Distribution<$ty> for Standard {
-            #[inline]
-            fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> $ty {
-                let mut vec: $ty = Default::default();
-                unsafe {
-                    let ptr = &mut vec;
-                    let b_ptr = &mut *(ptr as *mut $ty as *mut [u8; $bits/8]);
-                    rng.fill_bytes(b_ptr);
+// x86/64 are already little endian so we don't need packed_simd's `to_le` and
+// therefore can provide this on stable Rust.
+macro_rules! intrinsic_native_le_impl {
+    ($(($ty:ty, $init:ident)),+) => {
+        $(
+            impl Distribution<$ty> for Standard {
+                /// This is supported on x86/64 and supported target features only.
+                #[inline]
+                fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> $ty {
+                    let mut vec: $ty = unsafe { $init() };
+                    unsafe {
+                        let ptr = &mut vec;
+                        let b_ptr = &mut *(ptr as *mut $ty as *mut [u8; mem::size_of::<$ty>()]);
+                        rng.fill_bytes(b_ptr);
+                    }
+                    vec
                 }
-                vec.to_le()
             }
-        }
+        )+
     };
 }
 
-#[cfg(feature = "simd_support")]
-simd_impl!(16, u8x2, i8x2,);
-#[cfg(feature = "simd_support")]
-simd_impl!(32, u8x4, i8x4, u16x2, i16x2,);
-#[cfg(feature = "simd_support")]
-simd_impl!(64, u8x8, i8x8, u16x4, i16x4, u32x2, i32x2,);
-#[cfg(feature = "simd_support")]
-simd_impl!(128, u8x16, i8x16, u16x8, i16x8, u32x4, i32x4, u64x2, i64x2,);
-#[cfg(feature = "simd_support")]
-simd_impl!(256, u8x32, i8x32, u16x16, i16x16, u32x8, i32x8, u64x4, i64x4,);
-#[cfg(feature = "simd_support")]
-simd_impl!(512, u8x64, i8x64, u16x32, i16x32, u32x16, i32x16, u64x8, i64x8,);
-#[cfg(all(
-    feature = "simd_support",
-    any(target_arch = "x86", target_arch = "x86_64")
-))]
-simd_impl!((__m128i, u8x16), (__m256i, u8x32),);
+// this could perhaps be _mm_undefined_si128 but it seems the return type
+// for that will change to MaybeUninit<__m128i>
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+intrinsic_native_le_impl!(
+    (__m128i, _mm_setzero_si128),
+    (__m256i, _mm256_setzero_si256)
```

Comment:
I'm baffled: (1) the *types* exist without additional target features while the constructors require (`sse2` / `avx`), and (2) the constructors are *unsafe*. Maybe I should learn a little more about SIMD here...

Stupid questions, but:

1. This code will fail to compile without `sse2` / `avx`, right?
2. Is there a reason we shouldn't simply transmute an array with suitable alignment?

Reply:
AFAIK there are no dedicated instructions for the `setzero` intrinsics. Usually they get compiled either down to XORing the same register or to writing zero bytes to memory. I am also a bit surprised that they are gated on `sse2`/`avx`.

I agree that transmuting arrays would be a simpler solution, but instead of creating an array with proper alignment I think it will be easier to write something like this:
```rust
let mut buf = [0u8; mem::size_of::<$ty>()];
rng.fill_bytes(&mut buf);
unsafe {  mem::transmute_copy(&buf) }
```
`transmute_copy` will handle the alignment requirements and in practice should be properly optimized out by compiler.

Reply:
It will compile just fine but without see/avx it will fail to run

Reply:
@TheIronBorn 
Using intrinsics without properly checking required target features (either at compile or at run time) is considered UB.